pennylane/ftqc/graph_state_preparation.py

```diff
@@ -0,0 +1,166 @@
+
+r"""This module contains the GraphStatePrep template."""
+
+from typing import Union
+
+import networkx as nx
+
+import pennylane as qml
+from pennylane.operation import Operation
+from pennylane.wires import Wires
+
+from .qubit_graph import QubitGraph
+
+
+class GraphStatePrep(Operation):
+    r"""
+    Encode a graph state with the specified lattice structure, operations on each qubit, entanglement operations for nearest qubits and qubit graph.
+    The initial graph is :math:`|0\rangle^{V}`, given each qubit in the graph is in the :math:`|0\rangle` state and is not entangled with each other.
+    The target graph state :math:`| \psi \rangle = \prod\limits_{\{a, b\} \in E} U_{ab}|+\rangle^{V}` is prepared as below:
+    1. Each qubit is prepared as :math:`|+\rangle^{V}` state by applying the `qubit_ops` (`Hadamard` gate) operation.
+    2. Entangle every nearest qubit pair in the graph with `entanglement_ops` (`CZ` gate) operation.
+
+    Args:
+        graph (Union[QubitGraph, nx.Graph]): QubitGraph or nx.Graph object mapping qubit to wires.
+        qubit_ops (Operation): Operator to prepare the initial state of each qubit. Default as ``qml.H``.
+        entanglement_ops (Operation): Operator to entangle nearest qubits. Default as ``qml.CZ``.
+        wires: Wires the graph state preparation to apply on. Default as None.
+
+    **Example:**
+        The graph state preparation layer can be customized by the user.
+
+        .. code-block:: python3
+
+            import pennylane as qml
+            from pennylane.ftqc import generate_lattice, GraphStatePrep, QubitGraph
+
+            dev = qml.device('default.qubit')
+
+            @qml.qnode(dev)
+            def circuit(q, qubit_ops, entangle_ops):
+                GraphStatePrep(qubit_graph=q, qubit_ops=qubit_ops, entanglement_ops=entangle_ops)
+                return qml.probs()
+
+            lattice = generate_lattice([2, 2], "square")
+            q = QubitGraph("square", lattice.graph)
+
+            qubit_ops = qml.Y
+            entangle_ops = qml.CNOT
+
+        The resulting circuit after applying the ``GraphStatePrep`` template is:
+
+        >>> print(qml.draw(circuit, level="device")(q, qubit_ops, entangle_ops))
+        QubitGraph<square, (0, 0)>: ──Y─╭●─╭●───────┤  Probs
+        QubitGraph<square, (0, 1)>: ──Y─│──╰X─╭●────┤  Probs
+        QubitGraph<square, (1, 0)>: ──Y─╰X────│──╭●─┤  Probs
+        QubitGraph<square, (1, 1)>: ──Y───────╰X─╰X─┤  Probs
+    """
+
+    def __init__(
+        self,
+        graph: Union[nx.Graph, QubitGraph],
+        qubit_ops: Operation = qml.H,
+        entanglement_ops: Operation = qml.CZ,
+        wires: Wires = None,
+    ):
+        self.hyperparameters["graph"] = graph
+        self.hyperparameters["qubit_ops"] = qubit_ops
+        self.hyperparameters["entanglement_ops"] = entanglement_ops
+
+        if isinstance(graph, QubitGraph):
+            if wires is not None and set(wires) != set(graph.graph):
```

Comment:
[Optional] Converting a networkx graph to a set is equivalent to converting its nodes to a set (as I understand it), although it might be more expressive if we explicitly convert the QubitGraph's node labels to a set.

```suggestion
            if wires is not None and set(wires) != set(graph.node_labels):
```